src/mlpack/core/metrics/bleu.hpp

```diff
@@ -0,0 +1,165 @@
+   * \f$ reference_i = {paragraph_1, paragraph_2, ... } \f$
+   * and then each paragraph is an array of tokenized words/string. Like,
+   * \f$ paragraph_i = {word_1, word_2, ... } \f$.
+   * For ex.
+   * refCorpus = {{{"this", "is", "paragraph", "1", "from", "document", "1"},
+   *               {"this", "is", "paragraph", "2", "from", "document", "1"}},
+   *
+   *              {{"this", "is", "paragraph", "1", "from", "document", "2"},
+   *               {"this", "is", "paragraph", "2", "from", "document", "2"}}}
+   * @param translationCorpus It is an array of paragraphs which has been
+   * machine translated or generated for any natural language processing task.
+   * \f$ translationCorpus = {paragraph_1, paragraph_2, ... } \f$. So, each
+   * paragraph is an array of words. The ith paragraph from the corpus is
+   * \f$ paragraph_i = {word_1, word_2, ... } \f$.
+   * For ex.
+   * transCorpus = {{"this", "is", "generated", "paragraph", "1"},
+   *                {"this", "is", "generated", "paragraph", "2"}}
+   * @param smooth Whether or not to apply Lin et al. 2004 smoothing.
+   * @return The Evaluate method returns the BLEU Score. This method also
+   * returns other BLEU metrics (brevity penalty, translation length, reference
+   * length, ratio and precisions) which can be accessed by their corresponding
+   * accessor methods.
+   */
+  template <typename ReferenceCorpusType, typename TranslationCorpusType>
+  ElemType Evaluate(const ReferenceCorpusType& referenceCorpus,
+                    const TranslationCorpusType& translationCorpus,
+                    const bool smooth = false);
+
+  //! Serialize the metric (nothing to do).
+  template<typename Archive>
+  void serialize(Archive& /* ar */, const unsigned int /* version */) { }
+
+  //! Get the value of maximum length of tokens in n-grams.
+  size_t MaxOrder() const { return maxOrder; }
+  //! Modify the value of maximum length of tokens in n-grams.
+  size_t& MaxOrder() { return maxOrder; }
+
+  //! Get the BLEU Score.
+  ElemType BLEUScore() const { return bleuScore; }
+
+  //! Get the brevity penalty.
+  ElemType BrevityPenalty() const { return brevityPenalty; }
+
+  //! Get the value of translation length.
+  size_t TranslationLength() const { return translationLength; }
+
+  //! Get the value of reference length.
+  size_t ReferenceLength() const { return referenceLength; }
+
+  //! Get the ratio of translation to reference length ratio.
+  ElemType Ratio() const { return ratio; }
+
+  //! Get the precisions for corresponding order.
+  PrecisionType const& Precisions() const { return precisions; }
+
+ private:
+  /**
+   * Extracts all the n-grams.
+   *
+   * @tparam WordVector Type of the tokenized vector.
+   * @param segment Tokenized sequence represented in form of vector.
+   */
+  template <typename WordVector = std::vector<std::string>>
+  std::map<WordVector, size_t> GetNGrams(const WordVector& segment);
```

Comment:
I think you don't need `= std::vector<std::string>` here, the compiler should be able to detect it automatically.

Reply:
Right. Here it's not needed.

Reply:
Sure.